controllers/networkplugins_controller.go

```diff
@@ -17,35 +17,584 @@ limitations under the License.
 package controllers
 
 import (
+	"bytes"
 	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"reflect"
+	"text/template"
 
 	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/yaml"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	plumberv1 "github.com/platform9/luigi/api/v1"
+	"github.com/platform9/luigi/pkg/apply"
+)
+
+const (
+	DefaultNamespace        = "kube-system"
+	MultusImage             = "nfvpe/multus:v3.6"
+	WhereaboutsImage        = "xagent003/whereabouts:latest"
+	SriovCniImage           = "nfvpe/sriov-cni"
+	SriovDpImage            = "nfvpe/sriov-device-plugin:v3.2"
+	HostplumberImage        = "platform9/luigi-plumber:latest"
+	NfdImage                = "k8s.gcr.io/nfd/node-feature-discovery:v0.6.0"
+	TemplateDir             = "/etc/plugin_templates/"
+	CreateDir               = TemplateDir + "create/"
+	DeleteDir               = TemplateDir + "delete/"
+	NetworkPluginsConfigMap = "pf9-networkplugins-config"
 )
 
 // NetworkPluginsReconciler reconciles a NetworkPlugins object
 type NetworkPluginsReconciler struct {
 	client.Client
-	Log    logr.Logger
-	Scheme *runtime.Scheme
+	Log         logr.Logger
+	Scheme      *runtime.Scheme
+	currentSpec *plumberv1.NetworkPluginsSpec
+	prevSpec    *plumberv1.NetworkPluginsSpec
 }
 
-// +kubebuilder:rbac:groups=plumber.k8snetworking.pf9.io,resources=networkplugins,verbs=get;list;watch;create;update;patch;delete
-// +kubebuilder:rbac:groups=plumber.k8snetworking.pf9.io,resources=networkplugins/status,verbs=get;update;patch
+type MultusT plumberv1.Multus
+type WhereaboutsT plumberv1.Whereabouts
+type SriovT plumberv1.Sriov
+type HostPlumberT plumberv1.HostPlumber
+type NodeFeatureDiscoveryT plumberv1.NodeFeatureDiscovery
```

Comment:
I need to rethink this whole mess. I was originally thinking of something like (in pseudo-code):

for plugin in networkPlugins:
    plugin.ParseTemplate()
    plugin.ApplyTemplate()

So I made an interface, but maybe that's not the best design. I could just write explicit functions to handle each plugin, since I already have such if() checks in my code. And then move each plugin specific function under pkg/plugins/<name of plugin>

if networkPlugins.Multus != nil {
    parseMultusConfig(networkPlugins.Multus)
    applyMultusPlugins(networkPlugins.Multus)
}
... repeat for each defined CNI